PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+    # process distance matrix
+    # - this should be sorted (PyTorch allows this to be done on GPUs)
+    # - then the functions can be reimplemented to run faster on a
+    #   sorted list
+    distance_index = 1 if use_accessory else 0
+    distances = X[:,distance_index]
+    
+    # sort distances
+    distance_ranks = np.argsort(distances)
+    distances = distances[distance_ranks]
+    
+    # determine whether ref-ref or ref-query analysis
+    isolate_list = rlist
+    
+    # determine whether novel analysis or modifying existing analysis
+    use_existing = False
+    neighbours = {}
+    lineage_seed = {}
+    lineage_clustering = {}
+    previous_lineage_clustering = {}
+    
+    null_cluster_value = len(isolate_list) + 1
+    for R in rank_list:
+        lineage_clustering[R] = {i:null_cluster_value for i in isolate_list}
+        lineage_seed[R] = {}
+        neighbours[R] = {}
+        previous_lineage_clustering[R] = {}
+    
+    # shared memory data structures
+    shm_distances = shared_memory.SharedMemory(create = True, size = distances.nbytes, name = 'shm_distances')
+    distances_shared = np.ndarray(distances.shape, dtype = distances.dtype, buffer = shm_distances.buf)
+    distances_shared[:] = distances[:]
+
+    shm_distance_ranks = shared_memory.SharedMemory(create = True, size = distance_ranks.nbytes, name = 'shm_distance_ranks')
+    distance_ranks_shared = np.ndarray(distance_ranks.shape, dtype = distance_ranks.dtype, buffer = shm_distance_ranks.buf)
+    distance_ranks_shared[:] = distance_ranks[:]
+
+    isolate_list_shared = shared_memory.ShareableList(isolate_list, name = 'shm_isolate_list')
+    
+    # run clustering for an individual R
+    if num_processes == 1:
+        for R in rank_list:
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = run_clustering_for_R(R,
+                                                                                        null_cluster_value = null_cluster_value,
+                                                                                        qlist = qlist,
+                                                                                        existing_scheme = existing_scheme,
+                                                                                        distances_length = distances.shape,
+                                                                                        distances_type = distances.dtype)
+
+
+    else:
+
+        with Pool(processes = num_processes) as pool:
+            results = pool.map(partial(run_clustering_for_R,
+                                null_cluster_value = null_cluster_value,
+                                qlist = qlist,
+                                existing_scheme = existing_scheme,
+                                distances_length = distances.shape,
+                                distances_type = distances.dtype),
+                                rank_list)
```

Comment:
Likely that len(R) < n cores... speeding up and parallelising the sort using all cores may be more effective (if that's the slow step)

Reply:
Should I limit num processes to number of ranks or number of CPUs, whichever is lower?